File: 48_longest_no_repeat_substring/main.cpp
#include <iostream>
#include <vector>

int longestSubstringNoRepeat(const std::string& str) {
    int cur_len = 0;
    int max_len = 0;
    std::vector<int> pos(26);
    for (int i = 0; i < 26; ++i) pos[i] = -1;

    for (int i = 0; i < str.length(); ++i) {
        int prev_idx = pos[str[i] - 'a'];
        if (prev_idx < 0 || i - prev_idx > cur_len) ++cur_len;
        else {
            if (cur_len > max_len) {
                max_len = cur_len;
            }
            cur_len = i - prev_idx;
        }
        pos[str[i] - 'a'] = i;
    }
    if (cur_len > max_len) {
        max_len = cur_len;
    }

    return max_len;
}

int main() {
    std::cout << longestSubstringNoRepeat("arabcacfr") << std::endl;
}

File: 48_longest_no_repeat_substring/main.py
# -*- coding: utf-8 -*-


def longest_substring_no_repeat(s):
    cur_len, max_len, pos = 0, 0, [-1] * 26

    for i in range(len(s)):
        prev_idx = pos[ord(s[i]) - 97]
        if prev_idx < 0 or i - prev_idx > cur_len:
            cur_len += 1
        else:
            if cur_len > max_len:
                max_len = cur_len
            cur_len = i - prev_idx
        pos[ord(s[i]) - 97] = i
    if cur_len > max_len:
        max_len = cur_len
    return max_len


if __name__ == '__main__':
    print(longest_substring_no_repeat('arabcacfr'))
